android/app/src/main/java/com/microsoft/reacttestapp/MainActivity.kt

```diff
@@ -76,48 +76,32 @@ class MainActivity : ReactActivity() {
-                                ctx.runOnJSQueueThread {
-                                    val appKeys = AppRegistry.getAppKeys(ctx)
-                                    val viewModels = appKeys.map { appKey ->
-                                        ComponentViewModel(appKey, appKey, null, null)
-                                    }
-                                    mainThreadHandler.post {
-                                        componentListAdapter.setComponents(viewModels)
-                                        if (isTopResumedActivity && viewModels.count() == 1) {
-                                            startComponent(viewModels[0])
-                                        }
-                                    }
+                    testApp.reactNativeHost.addReactInstanceEventListener { context ->
+                        val ctx = context as ReactApplicationContext
+                        ctx.runOnJSQueueThread {
+                            val appKeys = AppRegistry.getAppKeys(ctx)
+                            val viewModels = appKeys.map { appKey ->
+                                ComponentViewModel(appKey, appKey, null, null)
+                            }
+                            mainThreadHandler.post {
+                                componentListAdapter.setComponents(viewModels)
+                                if (isTopResumedActivity && viewModels.count() == 1) {
+                                    startComponent(viewModels[0])
                                 }
                             }
                         }
-                    )
+                    }
                 } else {
-                    val index =
-                        if (components.count() == 1) 0 else session.lastOpenedComponent(checksum)
+                    val singleComponent = components.count() == 1
+                    val index = if (singleComponent) 0 else session.lastOpenedComponent(checksum)
                     index?.let {
-                        val component = newComponentViewModel(components[it])
-                        val startInitialComponent = object : ReactInstanceEventListener {
-                            override fun onReactContextInitialized(context: ReactContext) {
-                                if (isTopResumedActivity) {
-                                    startComponent(component)
-                                }
-                            }
-                        }
-                        testApp.reactNativeHost.apply {
-                            addReactInstanceEventListener(startInitialComponent)
-                            reactInstanceManager.currentReactContext?.let {
-                                startInitialComponent.onReactContextInitialized(it)
-                            }
-                        }
+                        val component = components[it]
+                        startComponent(newComponentViewModel(component))
```

Comment:
This is the only relevant change. The other stuff are just some minor refactoring.